### PR TITLE
armtz: switch to use OPTEE MSG protocol

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Switches to the new OPTEE_MSG protocol also used by the generic TEE
driver in secure world communication.

Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (QEMU)
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

Depends on https://github.com/OP-TEE/optee_client/pull/46

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Depends on https://github.com/OP-TEE/optee_os/pull/649

**Comment by jforissier** [comment]
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey)`

(requires https://github.com/OP-TEE/optee_linuxdriver/pull/44)

**Comment by jenswi-linaro** [comment]
Added the fix from Jerome

File: armtz/tee_tz_drv.c [hunk 1]
```diff
@@ -606,65 +604,48 @@ static int tz_open(struct tee_session *sess, struct tee_cmd *cmd)
...
-	arg32 = alloc_tee_arg(ptee, &parg32, TEESMC32_GET_ARG_SIZE(
-				TEEC_CONFIG_PAYLOAD_REF_COUNT + num_meta));
-	meta = alloc_tee_arg(ptee, &pmeta, sizeof(*meta));
+	arg = alloc_tee_arg(ptee, &parg, OPTEE_MSG_GET_ARG_SIZE(
+				TEEC_CONFIG_PAYLOAD_REF_COUNT + 2));
 
-	if ((arg32 == NULL) || (meta == NULL)) {
-		free_tee_arg(ptee, parg32);
-		free_tee_arg(ptee, pmeta);
+	if ((arg == NULL)) {
```
Comment (jforissier) [inline armtz/tee_tz_drv.c]: `if (!arg)`

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswi-linaro** [comment]
Rebased and reordered patches. Synced with review update from optee_os.

File: armtz/tee_tz_drv.c [hunk 2]
```diff
@@ -198,211 +171,242 @@ static void handle_rpc_func_cmd_wait(struct teesmc32_arg *arg32)
...
+	struct timespec ts;
+
+	if (arg->num_params != 1)
+		goto bad;
+	params = OPTEE_MSG_GET_PARAMS(arg);
+	if ((params->attr & OPTEE_MSG_ATTR_TYPE_MASK) !=
+			OPTEE_MSG_ATTR_TYPE_VALUE_OUTPUT)
+		goto bad;
+
+	getnstimeofday(&ts);
```
Comment (pascal-brand-st-dev) [inline armtz/tee_tz_drv.c]: Is it REE time?

**Comment by pascal-brand-st-dev** [comment]
Tested-by: Pascal Brand <pascal.brand@linaro.org> (STM)

File: armtz/tee_tz_drv.c (hunk 2 above)
Comment (jenswi-linaro) [inline armtz/tee_tz_drv.c]: Yes

**Comment by pascal-brand-st-dev** [comment]
Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

**Comment by jenswi-linaro** [comment]
Rebased, tags applied

**Pull request merged**